### DISCUSSION TITLE
Project Status

### DISCUSSION THREAD

**Discussion by gucci-on-fleek** [gucci-on-fleek/lockdown-browser#53] [Q&A]
(*TL;DR:* Respondus's *internal bug tracker has linked to this repository, suggesting that the* Lockdown Browser *will soon be updated to prevent this tool from working.*)

Project Status
==============

### Background

Back when <ac>COVID</ac> first hit, some of my professors began to mandate that we use _Respondus Lockdown Browser_ and _Respondus Monitor_ to write our exams. I was incensed that we were required to install this invasive software on our personal computers and submit to being recorded by some random company, with no option to opt-out. I argued with my professors that this requirement was both unnecessary and unethical, but they refused to budge.

I proceeded to submit [a formal complaint](https://github.com/gucci-on-fleek/lockdown-browser/files/14229105/respondus-concerns-redacted.pdf) to the administration, and after a protracted series of emails, they eventually offered an alternative, and having no other options, I accepted. But this alternate writing method was fairly arduous: before each exam, I would drive to the campus, check out a laptop, and drive back home. On my personal laptop, I would start a video call with my professor who would watch me while I wrote the exam; on the campus laptop, I would use the _Lockdown Browser_ (without _Monitor_) to take the exam. Once I was finished, I would drive back to campus to return the laptop. And furthermore, I was the only student who was allowed to use this method—everyone else was still required to use _Respondus Monitor_.

### Development

So then why did I decide to write this project?

1.  To show that the _Browser_ is ineffective.

    All of my complaints were ignored because the university considers “preventing cheating” to be more important than the privacy of its students. But if the _Browser_ were shown as being completely useless, then the university would have no reason to continue using it.

2.  Help honest students protect their privacy.

    Not all students were as lucky as I was to be offered an alternative writing method. I released this project to help the students who legitimately care about their privacy to partially alleviate some of the _Browser_'s flaws.

    Of course, it is certainly possible to use this project to cheat on exams, but you could say that about nearly any technology. The mere _possibility_ of cheating is not a valid reason to invade the privacy of every student.

### Present Day

Four years ago, I started this project. It has now been three years since I transferred to a different university that does not use any invasive monitoring software, and two years since I have had any access to a computer that runs _Windows_.

What does this mean for the project? It means that I have little motivation or ability to make any significant updates to it. I do try to reply to [issues](https://github.com/gucci-on-fleek/lockdown-browser/issues) and [discussions](https://github.com/gucci-on-fleek/lockdown-browser/discussions) whenever possible, but I've been quite slow at responding for this past year since I've been busy with other things. I tend to respond quicker to [pull requests](https://github.com/gucci-on-fleek/lockdown-browser/pulls) though.

### Future

So why am I posting this update now? Well, the project has recently jumped in popularity:

![Star Graph](https://api.star-history.com/svg?repos=gucci-on-fleek/lockdown-browser&theme=dark)

Which has likely lead to this:

![GitHub Traffic Screenshot](https://github.com/gucci-on-fleek/lockdown-browser/assets/49086429/fdeb35d0-7ddb-4852-bd16-6b1bb97a61ff)

[`respondus2.com`](https://archive.today/dpfo5#selection-309.166-309.180) is an official _Respondus_ domain, and [Trac](https://en.wikipedia.org/wiki/Trac) is bug tracking software. This repository being linked to from an internal _Respondus_ bug tracker suggests that **the _Browser_ will soon be updated to prevent this project from working.**

So what comes next? Well, it's up to you. If you're able to find a way to patch around the _Browser_'s upcoming update, then [submit a pull request](https://github.com/gucci-on-fleek/lockdown-browser/pulls). Otherwise, you'll need to wait for someone else to do so.

### Links

- [“Federal Judge: Invasive Online Proctoring "Room Scans" Are Unconstitutional”, EFF](https://www.eff.org/deeplinks/2022/08/federal-judge-invasive-online-proctoring-room-scans-are-also-unconstitutional)

- [“After Students Challenged Proctoring Software, French Court Slaps TestWe App With a Suspension”, EFF](https://www.eff.org/deeplinks/2023/03/after-students-challenged-proctoring-software-french-court-slaps-testwe-app)

- [“A Long Overdue Reckoning For Online Proctoring Companies May Finally Be Here”, EFF](https://www.eff.org/deeplinks/2021/06/long-overdue-reckoning-online-proctoring-companies-may-finally-be-here)

- [“The Security Failures of Online Exam Proctoring”, Schneier on Security](https://www.schneier.com/blog/archives/2020/11/the-security-failures-of-online-exam-proctoring.html)

- [“The extremely shady "educational integrity" industry”, Cory Doctorow](https://pluralistic.net/2022/02/16/unauthorized-paper/#cheating-anticheat)

- [“The pandemic showed remote proctoring to be worse than useless”, Cory Doctorow](https://pluralistic.net/2021/06/24/proctor-ology/#miseducation)

- [“Can You Trust Your Computer?”, Richard Stallman](https://www.gnu.org/philosophy/can-you-trust.html)

- [“The Right to Read”, Richard Stallman](https://www.gnu.org/philosophy/right-to-read.html)

**Comment by ericswpark** [comment]
@gucci-on-fleek have you ever looked into running LDB on something like QEMU or VirtualBox? I gave it a shot last weekend but failed, probably due to LDB recognizing the QEMU device names which apparently cannot be changed without applying patches to the QEMU source and recompiling. Seems like it should be doable though as one comment I came across said that LDB only checks for the device name stuff and what not, and doesn't go as deep as RDTSC checking as checked by pafish. Any idea if this will be more resilient against Respondus potentially patching it out down the line?

EDIT: this should also hopefully help with the alt-tab/minimize issues mentioned here: #54

**Reply by gucci-on-fleek** [reply]
> probably due to LDB recognizing the QEMU device names

Yeah, that's why I used the _Windows Sandbox_ in the first place, since it looks almost identical to a real installation.

> which apparently cannot be changed without applying patches to the QEMU source and recompiling

Try just deleting the driver files then removing the device from “Device Manager”. Deleting the driver files shouldn't affect any drivers that are already loaded, and removing the device from “Device Manager” shouldn't actually unload the drivers for anything important. That's kinda what I'm doing right now inside the _Sandbox_:

https://github.com/gucci-on-fleek/lockdown-browser/blob/7248e2999c70bd913345483873cfc8e623bc99a0/runtime_directory/sandbox_run.ps1#L17-L24

Take a snapshot first though, since the machine might not boot afterwards.

> without applying patches to the QEMU source and recompiling

Which might actually be really easy, depending on how good/awful their build tooling is. The only tricky thing would be if the Windows drivers are digitally signed, because then recompiling would be impossible.

> have you ever looked into running LDB on something like QEMU or VirtualBox

If I were to try running the _Browser_ again, I'd probably try running a regular Windows installation inside KVM/QEMU, then running the _Windows Sandbox_ inside of that. Nested virtualisation used to be pretty painful if not impossible, but I've heard that it works reasonably well on modern hardware. 

> Any idea if this will be more resilient against Respondus potentially patching it out down the line?

Probably less resilient, since VirtualBox and QEMU leave way more evidence than the _Windows Sandbox_ does. Starting with Windows 11, [the entire base Windows installation is virtualized by default](https://learn.microsoft.com/en-us/windows-hardware/design/device-experiences/oem-hvci-enablement), so a blind “is this system virtualized” check would be useless.

But possibly more resilient, since it's possible to modify the VirtualBox/QEMU source to directly rename any of the obvious references. It's not really possible to modify the Hyper-V/Windows Sandbox source.

**Reply by ericswpark** [reply]
@gucci-on-fleek so I realized that the first lockout only checks the device names, and managed to proceed further by just using regedit (see my blog post here: https://ericswpark.com/blog/2024/2024-02-20-i-hate-lockdown-browser/). However, the second lockout with the stupid "virtual moose" thing shows up during the exam, and I'm guessing it relies on some other VM detection technique. Maybe someone might find a workaround with the QEMU XML changes I documented but these ones are really basic and I have no idea how much more changes are required.

**Reply by gucci-on-fleek** [reply]
Huh, I've screwed with the _Lockdown Browser_ quite a bit and I've never seen that moose before. Must be a (somewhat) recent addition. I've also never seen it do any sort of detection after it's completed the startup process either, so I'm a little surprised. 

Provided that you have somewhat recent hardware, you can try running the Windows Sandbox _inside_ KVM/QEMU. Here's [a random guide on enabling nested virtualization](https://docs.fedoraproject.org/en-US/quick-docs/using-nested-virtualization-in-kvm/). I'm not sure whether running two VMs will cancel each other out or just make it more prominent, but it should be fairly easy to enable, so it's worth a try.

And if you're trying to figure out how the _Lockdown Browser_ is detecting that it's in a VM, I used [Process Monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) and [Detours `traceapi`](https://github.com/microsoft/Detours/tree/main/samples/traceapi). There's a lot of output though, so you kinda need to know what you're looking for first.

Another option would be to try a macOS VM. To make macOS boot on non-Apple hardware, you already have to trick it into thinking that it's booting on genuine Mac hardware, so in theory, there should be nothing for the _Lockdown Browser_ to detect. No experience with this, but Googling [`macos kvm`](https://www.google.com/search?q=macos+kvm) gives tons of results.

**Comment by carrie-leigh-221** [comment]
I'm not a programmer by any means, but love your project.  It took me a few hours to get up a running, but it has worked flawlessly for about a year now.  I don't have the ability to work around any future updates to LDB, but could potentially sponsor someone else's efforts should this current method fail.

**Reply by ericswpark** [reply]
Honestly, if you are willing to pay, I would recommend grabbing a used laptop from eBay or a used marketplace and using that for the semester. Any software-based workarounds run the risk of detection. Sure, it's a bit wasteful, but if your institution refuses to change policy then it's the only reliable option. Plus, you could even sell it after the semester is over.

Alternatively, try asking your instructor if the requirement can be removed. One of my professors actually got rid of the LDB requirement! You might be pleasantly surprised, just ask :)

**Reply by carrie-leigh-221** [reply]
I'm not following -- what would a used laptop from eBay solve?  I'm sorry, I feel like there's something very obvious that I'm missing here.

**Reply by ericswpark** [reply]
My suggestion was to use some sort of burner laptop as the primary way of running LDB. That way it stays separate from your main machine and can't touch your files or mess up your computer.

**Comment by Abdelmouzahir** [comment]
<img width="680" alt="Capture" src="https://github.com/gucci-on-fleek/lockdown-browser/assets/123598139/4c3ad63b-599d-4c60-b121-403d4196d4ad">

I got this error while building

**Reply by gucci-on-fleek** [reply]
Please [open a new issue for this](https://github.com/gucci-on-fleek/lockdown-browser/issues/new).

<sup>_(Although I would guess that the problem is that you didn't install the [Visual Studio C++ Tools](https://github.com/gucci-on-fleek/lockdown-browser#system-requirements) properly.)_</sup>

**Comment by kiwi08522** [comment]


**Reply by carrie-leigh-221** [reply]
I can confirm it's still working on my setup as of today.

**Reply by gucci-on-fleek** [reply]
> Is this solution still effective?

Probably. `trac.respondus2.com` aged out off the “Referring sites” list after 2 weeks, without any additional views ever showing up. There was [maybe a _Browser_ update](https://archive.ph/jBTsz) right after their referrer showed up, but it was [maybe well before this](https://archive.ph/4HUev). I haven't heard any complaints recently about it not working (well, aside from [the constant influx](https://github.com/gucci-on-fleek/lockdown-browser/issues) from people who don't follow the instructions properly), so it's probably still fine.

**Reply by FIM43-Redeye** [reply]
Honestly, while I'm quite incompetent myself, I would assume that it's entirely possible that using Windows Sandbox to do the magic could put Respondus directly at odds with Microsoft, and incurring their active wrath for bypassing their own security policies in a way far more obviously a rootkit could be far more trouble than it's worth.

**Reply by gucci-on-fleek** [reply]
@FIM43-Redeye

> I would assume that it's entirely possible that using Windows Sandbox to do the magic could put Respondus directly at odds with Microsoft, and incurring their active wrath for bypassing their own security policies in a way far more obviously a rootkit could be far more trouble than it's worth.

I doubt that Microsoft really cares what the _Lockdown Browser_ is trying to do. As long as the _Browser_ is just trying to detect if it's running inside the _Windows Sandbox_ or not, then as far as Microsoft is concerned, that's the program's business and not theirs. And if the _Browser_ somehow succeeded at escaping the _Sandbox_ (which it isn't even trying to do), then [they'd get $250 000](https://www.microsoft.com/en-us/msrc/bounty-hyper-v).

In general, I can think of 3 scenarios where Microsoft would care what your program does.

1. Using unsupported internal interfaces.

    Microsoft doesn't like it if you use undocumented internal Windows interfaces, and the _Browser_ definitely does this in a few places. But the worst that can come of this is that Raymond Chen makes fun of you ([1](https://devblogs.microsoft.com/oldnewthing/20031223-00/?p=41373), [2](https://devblogs.microsoft.com/oldnewthing/20190724-00/?p=102730), [3](https://devblogs.microsoft.com/oldnewthing/20230113-00/?p=107706)) and the interface is removed.

2. You're writing a kernel driver.

    If you're writing a kernel driver, then [you can't run it unless it's signed by Microsoft](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/kernel-mode-code-signing-policy--windows-vista-and-later-). For Microsoft to sign your driver, [you need to follow a very specific set of rules](https://learn.microsoft.com/en-us/windows-hardware/drivers/install/kernel-mode-code-signing-requirements--windows-vista-and-later-), and if you break the rules, then they'll refuse to sign your drivers and potentially revoke/disable your older ones. But the _Browser_ doesn't load any drivers (kernel or otherwise), so this doesn't apply.
    
3. You're writing malware.

    Windows Defender will remove any programs that it considers to be malware or a PUA (**p**otentially **u**nwanted **a**pplication). [Looking at the categories of malware and PUAs](https://learn.microsoft.com/en-us/microsoft-365/security/defender/criteria?view=o365-worldwide), the _Browser_ might fall under “lack of control” or “poor removal experience”. However, [Microsoft has their own locked-down browser](https://learn.microsoft.com/en-us/education/windows/take-tests-in-windows), and [given their history with suppressing alternative browsers](https://en.wikipedia.org/wiki/BrowserChoice.eu), they would be very reluctant to have Windows Defender remove the _Lockdown Browser_.

So as disappointing as it is, Microsoft is unlikely to block or otherwise inhibit the _Lockdown Browser_.

**Comment by FIM43-Redeye** [comment]
Hey, just so you all know, it happened. Respondus got around to it. No longer safe.

**Reply by carrie-leigh-221** [reply]
Still working on my setup.  
![image](https://github.com/gucci-on-fleek/lockdown-browser/assets/143338638/1b21419c-c2b6-4df5-8e5e-3543eb24bb6c)

**Reply by carrie-leigh-221** [reply]
I just completed a quiz in LDB about two minutes ago.  It worked as it always has for me.

**Reply by kiwi08522** [reply]
@carrie-leigh-221 What version LDB installer are you using?

**Reply by carrie-leigh-221** [reply]
Pardon me, I'm kind of dumb.  Does this answer your question? 
![image](https://github.com/gucci-on-fleek/lockdown-browser/assets/143338638/8c4cf44a-87be-4259-85bd-91f9b0a4864a)

**Reply by kiwi08522** [reply]
Yes it does, thank you. Your version is a bit more 'outdated' than mine, I'm on 2-1-2-02, and the person who posted this is on 2-1-2-04, so I'm wondering if updating to the newest version would make it not work anymore. I have no idea if the LDB installer pulls updates automatically, but if it doesn't, it should still be confirmed to work at 2-1-2-02 and would always work as long as you don't get a new installer.

Respondus says:
Version 2.1.2.04, 3-14-24
Improvements to security
Version 2.1.2.03, 2-16-24
Hotfix for a startup issue that was affecting some Windows computers
Version 2.1.2.02, 1-30-24
Added Canvas Classic support for ReadSpeaker TextAid
Improvements to security
Version 2.1.2.01, 12-19-23
Fixed a screen recording issue introduced in the previous release

It seems to me, since it's still working for you, that it will continue to work for anyone who has an installer before 2.1.2.02, which I am on. I am not sure about 2.1.2.03. But the Lockdown installations are linked to your school so it will be hard for anyone to get an installer that is on a specific version unless they can find someone at their school who will give them the file and hope that person has the installer still saved.

**Reply by carrie-leigh-221** [reply]
@kiwi08522 any idea how to make LDB not auto update?  If I run 2.1.2.02, it auto updates to 2.1.2.04, which detects the VM.

**Reply by bluesangheili** [reply]
The ban may be done by your institutional account, as I tried it on two entirely different computers. I am very screwed unless I figure out a way out of this.

**Reply by ThunderRuler** [reply]
> @kiwi08522 any idea how to make LDB not auto update? If I run 2.1.2.02, it auto updates to 2.1.2.04, which detects the VM.

I would also like to know

**Reply by big5saf** [reply]
it is nolonger safe they are banning us

**Comment by big5saf** [comment]
respondus got to it and now it doesn't work i got banned permanently.please not that this is linked to the school account

**Comment by justinjin2** [comment]
Does anyone who has this working able to help me to get mine work step by step? Sorry I'm just not very tech savvy. Send me your discord tag and we can talk (I can pay a bit too if needed)
